Operator developer guidance (#72470)
pytorch/pytorch: torch/csrc/jit/operator_upgraders/README.md

After you change to operator either the operator schema is BC-breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking iff they are used in TorchScript or mobile. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `

The steps to write upgrader:

> 1. [Build PyTorch from souce](https://github.com/pytorch/pytorch#from-source) and prepare a test model before making changes to the operator, following the process below. A test model before making the operator changes is needed to test the upgrader. Otherwise, after the change to operator, the new runtime will no longer be able to produce a model with the historic operator and can't test it anymore.

Reviewer: It's confusing for bullet points. The "big" first step of "build from sources" covers "small" steps of adding testing modules. It does not make sense to me. Some reformatting may be needed.

>      1. Add a test module in `test/jit/fixtures_srcs/fixtures_src.py`. In `test/jit/fixtures_srcs/generate_models.py`, add a module

Reviewer: ```suggestion
     1. Add a test module in `test/jit/fixtures_srcs/fixtures_src.py`. In `test/jit/fixtures_srcs/generate_models.py`, 
```

# Guidance for Operator Developer

> PyTorch’s operators sometimes require changes to maintain the high quality user experience (UX) that PyTorch is known for. These changes can be backward compatibility (BC) breaking, where older programs will no longer run as expected on the latest version of PyTorch (an old writer / new reader problem) or forward compatibility (FC) breaking, where new programs will not run on older versions of PyTorch (a new writer / old reader problem). An upgrader is a method to use the new operator to mimic the old operator behavior. When a new runtime loads an old model with the old operator, the upgrader will replace the old operator in the model with the new operator. The replacement will only happen for old models, and it does not need to consider the new models. Please refer to the documentation [PyTorch Operator Versioning](https://github.com/pytorch/rfcs/blob/master/RFC-0017-PyTorch-Operator-Versioning.md) for more details.

Reviewer 2: A few things:
* I would rephrase old write / new reader to actually talk about program and runtime.
* We talk about FC but we should clarify upgraders are just for BC>
*  "The replacement will only happen for old models, and it does not need to consider the new models." => Shouldn't this just be as simple as "upgraders only apply to instances of old operators to make sure they comply with the new operator contract"?

Author: Address on the new pr

> After you change to operator either the operator schema is BC-breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking iff they are used in TorchScript or mobile. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `

Reviewer 2: There's typos here: "After you change to operator either the operator schema is BC-breaking way or the semantics of the operator"

Also, this should at least include some examples on what's BC breaking.

"change non-BC breaking iff they are used in TorchScript or mobile"
Just for mobile? Also how could anybody know that: they can't so this is not optional.

Author: Address on the new pr

> # Guidance for Operator Developer

Reviewer 3: I think Tugsuu is also adding this file in his PR:
https://github.com/pytorch/pytorch/pull/72313/files

Could you coordinate with him to land it? Thanks

Author: Yeah @tugsbayasgalan  reached out to me and mentioned about it. Sorry should've checked with him so there isn't overlap. I will address the comments from you and update this pr.

Author: https://github.com/pytorch/pytorch/pull/72313 includes code changes to skip bc check for operators with valid upgraders, as well as the guidance. It's probably better to have one dedicated pr for the guidance. I've added the changes from that pr to here and will address the feedback too.

Reviewer 4: Yeah. I was updating our FC/BC checking test and I needed to reference this doc so I included my own version in my PR. But I think it makes sense to land this PR first

> # Guidance for Operator Developer

Reviewer 2: We should have an intro on what the base concepts are: operator upgrader, BC and BC breaking, ...
Keep it short and simple.
You may already have something in the design review.

Author: Added

After you change either the operator signature in a BC breaking way or the semantics of the operator, you will need to write an “upgrader” to make the change non-BC breaking iff they are used in TorchScript or mobile. In general, you can know your operator is BC breaking, if it fails `test/forward_backward_compatibility/check_forward_backward_compatibility.py `

The steps to write upgrader:

> 1. Prepare a model before making changes to the operator, following the process:

Reviewer 2: Here you mean a test model, right?
Otherwise it's not clear what this model is mean to do or why. Please briefly explain.

Author: Add the reason why a test model is needed.

>      1. add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that, an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following

Reviewer 2: ```suggestion
     1. Add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following
```

>      1. add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that, an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following

Reviewer 2: what does this mean?

Author: Add more explanation and hope it's clearer.

>      1. add a module in `test/jit/fixtures_srcs/fixtures_src.py`. The reason why switching to commit is that, an old model with the old operator before the change is needed to ensure the upgrader is working as expected. In `test/jit/fixtures_srcs/generate_models.py`, add the module and its corresponding changed operator like following

Reviewer 2: probably should be another step

        ```
        ALL_MODULES = {
           TestVersionedLinspaceV7(): "aten::linspace",
        }

>         ```

Reviewer 2: Fix formatting for code blocks.

        This module should include the changed operator. If the operator isn't covered in the model, the model export process will fail.

       2. Export the model to `test/jit/fixtures` by running
        ```
        python test/jit/fixtures_src/generate_models.py
        ```

>        3. Commit the change and submit a pr

Reviewer 2: ```suggestion
       3. Commit the change and submit a pull request.
```

2. Make changes to the operator and write an upgrader.
    1. Make the operator change.

>     2. Write an upgrader in `caffe2/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp` file inside a map `kUpgradersEntryMap`. The softly enforced naming format is `&lt;operator_name>_&lt;operator_overload>_&lt;start>_&lt;end>`. For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need to be replaced by these upgraders.

Reviewer 2: will this render ok?

Author: Resolve

>     2. Write an upgrader in `caffe2/torch/csrc/jit/operator_upgraders/upgraders_entry.cpp` file inside a map `kUpgradersEntryMap`. The softly enforced naming format is `&lt;operator_name>_&lt;operator_overload>_&lt;start>_&lt;end>`. For example, the below example means that `linspace.out `and` linspace` at versions from 0 to 7 need to be replaced by these upgraders.

Reviewer 4: Maybe we should explain what version is? Because for general public, it is not very clear

Author: Add more explanation.

    ```
    static std::unordered_map<std::string, std::string> kUpgradersEntryMap(
        {
          {"linspace_0_7", R"SCRIPT(
    def linspace_0_7(start: Union[int, float, complex], end: Union[int, float, complex], steps: Optional[int], *, dtype: Optional[int], layout: Optional[int],
                     device: Optional[Device], pin_memory: Optional[bool]):
      if (steps is None):
        return torch.linspace(start=start, end=end, steps=100, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory)
      return torch.linspace(start=start, end=end, steps=steps, dtype=dtype, layout=layout, device=device, pin_memory=pin_memory)
    )SCRIPT"},
        }
    ```

>     2. Bump the `kProducedFileFormatVersion` by 1 and provide the reasons under `caffe2/caffe2/versions.h `

Reviewer 4: I think you also need to bump `kMaxProducedFileFormatVersion`

Author: Resolve

    ```
    // We describe new operator version bump reasons here:
    // 1) [01/24/2022]
    //     We bump the version number to 8 to update aten::linspace
    //     and aten::linspace.out to error out when steps is not
    //     provided. (see: https://github.com/pytorch/pytorch/issues/55951)
    // 2) [01/30/2022]
    //     Bump the version number to 9 to update aten::logspace and
    //     and aten::logspace.out to error out when steps is not
    //     provided. (see: https://github.com/pytorch/pytorch/issues/55951)
    constexpr uint64_t kProducedFileFormatVersion = 0x9L;
    ```

>     3. In `caffe2/torch/csrc/jit/operator_upgraders/version_map.cpp`, add changes like below. You will need to make sure that the entry is **SORTED** according to the version bump number.

Reviewer 4: Maybe mention how the below example is formatted according to what we describe here

Author: Added an example to explain

    ```
    {{"aten::linspace",
      {{8,
        "linspace_0_7",
        "aten::linspace(Scalar start, Scalar end, int? steps=None, *, ScalarType? dtype=None, Layout? layout=None, Device? device=None, bool? pin_memory=None) -> Tensor"}}},
    ```

>     4. After rebuild PyTorch, run the following command and it will auto generate a change to `caffe2/torch/csrc/jit/mobile/upgrader_mobile.cpp`. After rebuild PyTorch from source (`python setup.py`), run

Author: 1. "auto generate a change to upgrader_mobile.cpp" -> "auto update the file upgrader_mobile.cpp"
2. Add the link to build pytorch
3. Add the first time to build in step 1

    ```
    python pytorch/tools/codegen/operator_versions/gen_mobile_upgraders.py
    ```

>     5. Add test:

Reviewer 2: Is this a test or tests?
